docs/protocol-initialization-facet.md

```diff
@@ -0,0 +1,57 @@
+[![banner](images/banner.png)](https://bosonprotocol.io)
+
+<h1 align="center">Boson Protocol V2</h1>
+
+### [Intro](../README.md) | [Audits](audits.md) | [Setup](setup.md) | [Tasks](tasks.md) | [Architecture](architecture.md) | [Domain Model](domain.md) | [State Machines](state-machines.md) | [Sequences](sequences.md)
+
+## Protocol initialization handler facet
+
+Protocol initialization handler facet has a special role in the whole [architecture](architecture.md). On one side we have Diamond specific facets - Loupe and Cut - which enable lookups and facet management. On the other hand we have Protocol specific facets which implements all features needed for protocol to work. Most of these facets needs to be initialized at the time when cuts are done, which among other includes registering its interface in the diamond. However in some cases initialization of a facet if different depending on whether whole protocol is deployed for the first time or it is just upgraded. Moreover, some upgrades might introduce a logic that relies on data that was not produced in old versions of the protocol. For that purpose we introduce Protocol initialization handler facet, which conceptually belongs between Diamond specific and Protocol specific facets. Its role is to enable smooth atomic upgrades and if necessary populate some of contract storage.
+
+Protocol initialization handler facets responsibilities are:
+- store version of the protocol,
+- forward initialization calls to individual facets,
+- if called during an upgrade, properly handle initialization data,
+- remove supported interfaces,
+- add supported interfaces.
+
+Two biggest benefits of the Protocol initialization handler are:
+- It allows atomic cuts of multiple facets, so there is generally no need to pause the protocol during the upgrade.
+- It allows custom storage manipulation, so if any back filling is needed, we have a way to do it.
+
+### Versioning
+
+Initialize accepts version encoded in bytes32. Every time upgrade takes place, an unique version must be supplied, otherwise initialization reverts. Depending on version, some additional checks can be done, for example upgrade to certain versions can only be done if current protocol version is exactly one version lower. This kind of restriction is planned to be used on every upgrade. This means that skipping versions during the upgrade will not be possible. If the latest version of protocol code is more than one version higher that current deployed protocol, all intermediate upgrades need to be done to upgrade to the lates version.
+
+Facets allows to query the current version by calling `getVersion`.
+
+### Facet initialization
+
+Each Protocol facet should have initialization function (even if no-op) to ensure consistency across codebase and different version. Initialization function should be written as if protocol is deployed for the first time (i.e. not taking into account potential effects on the upgrade).
```

Comment:
Could you please also add that `initializer` functions should NEVER be added to diamond functions list?